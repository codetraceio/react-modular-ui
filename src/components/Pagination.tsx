import React, { MouseEvent, useCallback, useContext, useMemo } from "react";

import { className } from "../utils/className";
import Icon from "./Icon";
import { ThemeContext } from "./ThemeContext";

export interface PaginationProps {
  size?: string | number;
  color?: string;
  count?: number;
  limit?: number;
  offset?: number;
  hideLastPage?: boolean;
  theme?: string;

  onChange?: (offset: number, page: number) => void;
}

export default function Pagination({
  size = 24,
  color,
  count = 0,
  limit = 10,
  offset,
  hideLastPage,
  theme,

  onChange,
  ...props
}: PaginationProps) {
  const themeContext = useContext(ThemeContext);

  const normalizedOffset = Math.max(Math.min(offset ?? 0, count - 1), 0);
  const page = limit > 0 ? Math.ceil(normalizedOffset / limit) + 1 : 0;
  const lastPage = limit > 0 ? Math.ceil(count / limit) : 0;
  const startPage = Math.max(page - 2, 1);
  const endPage = Math.min(page + 2, lastPage);

  const pages = useMemo(() => {
    const result = [];
    result.push(1);
    for (let i = startPage; i <= endPage; i++) {
      if (i !== 1 && i !== lastPage) {
        result.push(i);
      }
    }
    if (lastPage !== 1) {
      result.push(lastPage);
    }
    return result;
  }, [startPage, endPage, lastPage]);

  const handleChange = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => {
      const newPage = Number(event.currentTarget.getAttribute("data-page"));
      const newOffset: number = (newPage - 1) * limit;
      if (typeof onChange === "function") {
        onChange(newOffset, newPage);
      }
    },
    [limit, onChange],
  );

  const prevElement = useMemo(() => {
    return (
      <button
        className={className("pagination", "prev")}
        aria-disabled={page === 1}
        data-page={page - 1}
        onClick={handleChange}
      >
        <Icon height={size} icon="pagination" />
      </button>
    );
  }, [page, size, handleChange]);

  const nextElement = useMemo(() => {
    return (
      <button
        className={className("pagination", "next")}
        aria-disabled={page === lastPage}
        data-page={page + 1}
        onClick={handleChange}
      >
        <Icon height={size} icon="pagination" rotate={180} />
      </button>
    );
  }, [page, size, lastPage, handleChange]);

  const pagesElement = useMemo(() => {
    let prevPage = 0;
    const result: JSX.Element[] = [];
    pages.forEach((currentPage, index) => {
      let afterEllipsis = false;
      if (currentPage !== prevPage + 1) {
        result.push(
          <div
            key={"ellipsis." + currentPage}
            className={className("pagination", "ellipsis")}
          >
            â€¦
          </div>,
        );
        afterEllipsis = true;
      }

      if (!(index === pages.length - 1 && afterEllipsis && hideLastPage)) {
        const characterLength: number = currentPage.toString().length;
        let characters = "single";
        if (characterLength === 2) {
          characters = "double";
        }
        if (characterLength > 2) {
          characters = "multiple";
        }
        result.push(
          <button
            key={currentPage}
            className={className("pagination", "item")}
            data-current={page === currentPage}
            data-characters={characters}
            data-page={currentPage}
            onClick={handleChange}
          >
            {currentPage}
          </button>,
        );
      }
      prevPage = currentPage;
    });
    return result;
  }, [pages, page, hideLastPage, handleChange]);

  if (!page || count <= limit) {
    return null;
  }

  return (
    <div
      className={className("pagination")}
      data-size={size}
      data-color={color}
      data-theme={theme ?? themeContext}
      {...props}
    >
      {prevElement}
      {pagesElement}
      {nextElement}
    </div>
  );
}
