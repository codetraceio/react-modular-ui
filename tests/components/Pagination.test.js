import * as React from "react";
import * as renderer from "react-test-renderer";

import Pagination from "../../src/components/Pagination";

describe("Pagination", () => {
  it("should render Pagination", () => {
    const tree = renderer.create(
      <Pagination
        size="32"
        color="primary"
        count="100"
        limit="10"
        offset="20"
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
