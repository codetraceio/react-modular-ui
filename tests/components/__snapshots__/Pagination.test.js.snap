// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render Pagination 1`] = `
<div
  className="c-pagination -size-32 -color-primary"
>
  <div
    className="c-pagination--prev"
    onClick={[Function]}
  >
    <div
      className="c-icon -height-32"
    />
  </div>
  <div
    className="c-pagination--item -characters-single"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="c-pagination--item -characters-single"
    onClick={[Function]}
  >
    2
  </div>
  <div
    className="c-pagination--item -current -characters-single"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="c-pagination--item -characters-single"
    onClick={[Function]}
  >
    4
  </div>
  <div
    className="c-pagination--item -characters-single"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="c-pagination--ellipsis"
  >
    â€¦
  </div>
  <div
    className="c-pagination--item -characters-double"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="c-pagination--next"
    onClick={[Function]}
  >
    <div
      className="c-icon -height-32 -rotate-180"
    />
  </div>
</div>
`;
